typescript/sdk/src/ism/EvmIsmModule.hardhat-test.ts

```diff
@@ -0,0 +1,720 @@
+      const matches = await ismModuleMatchesConfig({
+        ism,
+        config,
+        configured: false,
+      });
+      expect(matches).to.be.true;
+    });
+
+    for (const type of [IsmType.ROUTING, IsmType.FALLBACK_ROUTING]) {
+      it(`deploys ${type} routingIsm with correct routes`, async () => {
+        exampleRoutingConfig.type = type as
+          | IsmType.ROUTING
+          | IsmType.FALLBACK_ROUTING;
+
+        const ism = await createIsmModule({ config: exampleRoutingConfig });
+        const matches = await ismModuleMatchesConfig({
+          ism,
+          config: exampleRoutingConfig,
+          configured: false,
+        });
+        expect(matches).to.be.true;
+      });
+    }
+
+    for (let i = 0; i < 16; i++) {
+      it(`deploys a random ism config #${i}`, async () => {
+        const config = randomIsmConfig();
+        let ism: EvmIsmModule;
+        try {
+          ism = await createIsmModule({ config });
+        } catch (e) {
+          console.error('Failed to deploy random ism config', e);
+          console.error(stringifyObject(config as object, 'json', 2));
+          process.exit(1);
+        }
+
+        try {
+          const matches = await ismModuleMatchesConfig({
+            ism,
+            config,
+            configured: false,
+          });
+          expect(matches).to.be.true;
+        } catch (e) {
+          console.error('Failed to match random ism config', e);
+          console.error('EXPECTED');
+          console.error(stringifyObject(config as object, 'yaml', 2));
+          console.error('ACTUAL');
+          console.error(
+            stringifyObject((await ism.read()) as object, 'yaml', 2),
+          );
+          process.exit(1);
+        }
+      });
+    }
+  });
+
+  describe('update', async () => {
+    for (const type of [IsmType.ROUTING, IsmType.FALLBACK_ROUTING]) {
+      it(`should skip deployment with warning if no chain metadata configured ${type}`, async () => {
+        exampleRoutingConfig.type = type as
+          | IsmType.ROUTING
+          | IsmType.FALLBACK_ROUTING;
+        exampleRoutingConfig.domains[TestChainName.test4] = {
+          type: IsmType.MESSAGE_ID_MULTISIG,
+          threshold: 1,
+          validators: [randomAddress()],
+        };
+
+        // create a new ISM
+        const ism = await createIsmModule({ config: exampleRoutingConfig });
+        const matches = await ismModuleMatchesConfig({
+          ism,
+          config: exampleRoutingConfig,
+          configured: false,
+        });
+        expect(matches).to.be.true;
+
+        // apply the config to the new ISM
+        const applyTxs = await ism.update(exampleRoutingConfig);
+        await logAndSendTransactions(applyTxs);
+
+        // add config for a domain the multiprovider doesn't have
+        exampleRoutingConfig.domains['test5'] = {
+          type: IsmType.MESSAGE_ID_MULTISIG,
+          threshold: 1,
+          validators: [randomAddress()],
+        };
```

Comment:
imo we should still deploy ISMs even if the multi provider doesnt have the domain?

Reply:
i suppose we could, makes the ISM part of this https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3836 easier